### PR TITLE
Add 'set src <ipv6>' functionality for route-maps

### PR DESCRIPTION
**Pull request by pavel-shirshov**
Also was added two commands:
show ipv6 protocol
ipv6 protocol route-map

### CODE REVIEW

File: lib/if.h [hunk 1]
```diff
@@ -244,6 +244,7 @@ extern int if_cmp_func (struct interface *, struct interface *);
 extern struct interface *if_create (const char *name, int namelen);
 extern struct interface *if_lookup_by_index (unsigned int);
 extern struct interface *if_lookup_exact_address (struct in_addr);
+extern struct interface *if_lookup_exact_address6 (struct in6_addr src);
```
Comment (qiluo-msft) [inline lib/if.h]: > src [](start = 66, length = 4)

remove parameter name as others. #Closed

File: zebra/zebra_routemap.c [hunk 2]
```diff
@@ -359,26 +359,46 @@ ALIAS (no_match_ip_address_prefix_list,
 
 DEFUN (set_src,
        set_src_cmd,
-       "set src A.B.C.D",
+       "set src (A.B.C.D|X:X::X:X)",
```
Comment (qiluo-msft) [inline zebra/zebra_routemap.c:362]: > X:X::X:X [](start = 25, length = 8)

A:B:C:D:E:F:G:H ? #Closed

File: zebra/zebra_routemap.c [hunk 3]
```diff
@@ -359,26 +359,46 @@ ALIAS (no_match_ip_address_prefix_list,
...
        SET_STR
        "src address for route\n"
        "src address\n")
 {
   struct in_addr src;
+  struct in6_addr src6;
   struct interface *pif;
+  int is_ipv4;
```
Comment (qiluo-msft) [inline zebra/zebra_routemap.c:370]: > int [](start = 2, length = 3)

bool ? #Closed

File: zebra/zebra_routemap.c [hunk 4]
```diff
@@ -359,26 +359,46 @@ ALIAS (no_match_ip_address_prefix_list,
...
+  int is_ipv6;
 
-  if (inet_pton(AF_INET, argv[0], &src) <= 0)
+  is_ipv4 = inet_pton(AF_INET, argv[0], &src) > 0 ? 1 : 0;
```
Comment (qiluo-msft) [inline zebra/zebra_routemap.c]: > ? 1 : 0 [](start = 50, length = 7)

remove " ? 1 : 0" #Closed

File: zebra/zebra_routemap.c [hunk 5]
```diff
@@ -359,26 +359,46 @@ ALIAS (no_match_ip_address_prefix_list,
...
       vty_out (vty, "%% not a local address%s", VTY_NEWLINE);
       return CMD_WARNING;
     }
 
-    pif = if_lookup_exact_address (src);
-    if (!pif)
+  if (is_ipv4)
+    {
+      pif = if_lookup_exact_address (src);
+      if (!pif)
```
Comment (qiluo-msft) [inline zebra/zebra_routemap.c]: > if (!pif) [](start = 6, length = 9)

Move outside the outer if block? #Resolved

File: zebra/zebra_routemap.c [hunk 6]
```diff
@@ -397,7 +417,7 @@ DEFUN (no_set_src,
 
 ALIAS (no_set_src,
        no_set_src_val_cmd,
-       "no set src (A.B.C.D)",
+       "no set src (A.B.C.D|X:X::X:X)",
```
Comment (qiluo-msft) [inline zebra/zebra_routemap.c:412]: > X:X::X:X [](start = 28, length = 8)

The same #Closed

File: zebra/zebra_rib.c [hunk 7]
```diff
@@ -365,13 +365,15 @@ nexthop_ipv4_ifindex_add (struct rib *rib, struct in_addr *ipv4,
...
 struct nexthop *
-nexthop_ipv6_add (struct rib *rib, struct in6_addr *ipv6)
+nexthop_ipv6_add (struct rib *rib, struct in6_addr *ipv6, struct in6_addr *src)
 {
   struct nexthop *nexthop;
 
   nexthop = XCALLOC (MTYPE_NEXTHOP, sizeof (struct nexthop));
   nexthop->type = NEXTHOP_TYPE_IPV6;
   nexthop->gate.ipv6 = *ipv6;
+  if (src)
```
Comment (qiluo-msft) [inline zebra/zebra_rib.c:375]: > if (src) [](start = 2, length = 8)

Remove the condition, assign ipv6 anytime. #WontFix

**Review by qiluo-msft — CHANGES_REQUESTED** [review]
As comments

File: zebra/zebra_routemap.c (hunk 2 above)
Comment (pavel-shirshov) [reply zebra/zebra_routemap.c:362]: It's a special parameter inside of quagga. Check https://github.com/Azure/sonic-quagga/blob/debian/0.99.24.1/lib/command.h#L264 #Closed

File: zebra/zebra_routemap.c (hunk 3 above)
Comment (pavel-shirshov) [reply zebra/zebra_routemap.c:370]: There're no bool type in the whole quagga codebase.
So I decided use their conventions. #Closed

File: zebra/zebra_routemap.c (hunk 6 above)
Comment (pavel-shirshov) [reply zebra/zebra_routemap.c:412]: It's a special parameter inside of quagga. Check https://github.com/Azure/sonic-quagga/blob/debian/0.99.24.1/lib/command.h#L264 #Closed

File: zebra/zebra_routemap.c (hunk 5 above)
Comment (pavel-shirshov) [reply zebra/zebra_routemap.c]: can you please explain it more? #Resolved

File: zebra/zebra_rib.c (hunk 7 above)
Comment (pavel-shirshov) [reply zebra/zebra_rib.c:375]: In this case the following line will cause a crash
https://github.com/Azure/sonic-quagga/blob/pavelsh/src/zebra/zebra_rib.c#L2963
 #Closed

File: zebra/zebra_routemap.c (hunk 5 above)
Comment (qiluo-msft) [reply zebra/zebra_routemap.c]: if (is_ipv4) pif = ...
else pif = ...

if (!pif)
{
}

---

In reply to: [85822103](https://github.com/Azure/sonic-quagga/pull/4#discussion_r85822103) [](ancestors = 85822103)

**Review by qiluo-msft — CHANGES_REQUESTED** [review]
as comments.

File: zebra/zebra_rib.c (hunk 7 above)
Comment (pavel-shirshov) [reply zebra/zebra_rib.c:375]: In this case the following line will cause a crash
[https://github.com/Azure/sonic-quagga/blob/pavelsh/src/zebra/zebra_rib.c#L2963](https://github.com/Azure/sonic-quagga/blob/pavelsh/src/zebra/zebra_rib.c#L2963)

---

In reply to: [85613447](https://github.com/Azure/sonic-quagga/pull/4#discussion_r85613447) [](ancestors = 85613447)

Comment (qiluo-msft) [reply zebra/zebra_rib.c:375]: I see. Then
if (src == NULL) nexthop->src.ipv6 = all_zero_address
 #WontFix

Comment (pavel-shirshov) [reply zebra/zebra_rib.c:375]: Check this function from quagga nexthop_ipv4_ifindex_add(). I copied my ipv6 function from this.

---

In reply to: [86056451](https://github.com/Azure/sonic-quagga/pull/4#discussion_r86056451) [](ancestors = 86056451)

File: zebra/rt_netlink.c [hunk 8]
```diff
@@ -1805,6 +1817,7 @@ netlink_route_multipath (int cmd, struct prefix *p, struct rib *rib,
                 SET_FLAG (nexthop->flags, NEXTHOP_FLAG_FIB);
             }
         }
+
       if (src)
         addattr_l (&req.n, sizeof req, RTA_PREFSRC, &src->ipv4, bytelen);
```
Comment (lguohan) [inline zebra/rt_netlink.c:1822]: is this correct? only for ipv4? I noticed that you have modified the build_multipath for v6, but however, I feel ipv6 is missing here. #Resolved

Comment (pavel-shirshov) [reply zebra/rt_netlink.c:1822]: This code should work only for multipath case.
I don't have anything related ipv6 multipath in the codebase yet.
But I could add this into the code. It doesn't matter, honestly.

---

In reply to: [86168202](https://github.com/Azure/sonic-quagga/pull/4#discussion_r86168202) [](ancestors = 86168202)

**Review by lguohan — COMMENTED** [review]
I raised one question w.r.t. to the src.

**Review by lguohan — APPROVED** [review]

**Pull request merged**